File: front/central/src/services/auth/permissions/domain/ports.ts
import {
    Permission,
    PaginatedResponse,
    GetPermissionsParams,
    SingleResponse,
    CreatePermissionDTO,
    UpdatePermissionDTO,
    ActionResponse,
    BulkCreatePermissionsResponse
} from './types';

export interface IPermissionRepository {
    getPermissions(params?: GetPermissionsParams): Promise<PaginatedResponse<Permission>>;
    getPermissionById(id: number): Promise<SingleResponse<Permission>>;
    getPermissionsByScope(scopeId: number): Promise<PaginatedResponse<Permission>>;
    getPermissionsByResource(resource: string): Promise<PaginatedResponse<Permission>>;
    createPermission(data: CreatePermissionDTO): Promise<SingleResponse<Permission>>;
    updatePermission(id: number, data: UpdatePermissionDTO): Promise<SingleResponse<Permission>>;
    deletePermission(id: number): Promise<ActionResponse>;
    createPermissionsBulk(permissions: CreatePermissionDTO[]): Promise<BulkCreatePermissionsResponse>;
}

File: back/central/services/auth/permissions/internal/domain/ports.go
package domain

import "context"

type IPermissionRepository interface {
	GetPermissions(ctx context.Context, businessTypeID *uint, name *string, scopeID *uint, resource *string) ([]Permission, error)
	GetPermissionByID(ctx context.Context, id uint) (*Permission, error)
	GetPermissionsByScopeID(ctx context.Context, scopeID uint) ([]Permission, error)
	GetPermissionsByResource(ctx context.Context, resource string) ([]Permission, error)
	PermissionExistsByName(ctx context.Context, name string) (bool, error)
	CreatePermission(ctx context.Context, permission Permission) (string, error)
	UpdatePermission(ctx context.Context, id uint, permission Permission) (string, error)
	DeletePermission(ctx context.Context, id uint) (string, error)
}
